### ISSUE TITLE
bool operation difference panics for two polygons

### ISSUE THREAD

**Issue by Madoshakalaka** [georust/geo#865]
when differencing two polygons, geo panics with:
```
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `Clockwise`,
 right: `CounterClockwise`', /home/maa/.cargo/registry/src/github.com-1ecc6299db9ec823/geo-0.22.0/src/algorithm/bool_ops/rings.rs:162:13
```
I notice this happens when the inner polygon has a long and narrow slit.

I logged and visualized an incidence, both polygons only have exterior lines:
The exterior of the outer rectangle:
```
Coordinate { x: -640.0, y: -360.0 }, Coordinate { x: 640.0, y: -360.0 }, Coordinate { x: 640.0, y: 360.0 }, Coordinate { x: -640.0, y: 360.0 }, Coordinate { x: -640.0, y: -360.0 } 
```
The exterior of the inner irregular polygon:
```
Coordinate { x: 313.276, y: 359.999 }, Coordinate { x: 213.319, y: 359.999 }, Coordinate { x: 50.0, y: 60.0 }, Coordinate { x: -50.0, y: 60.0 }, Coordinate { x: -50.0, y: 110.0 }, Coordinate { x: -8.817, y: 360.0 }, Coordinate { x: -93.151, y: 360.0 }, Coordinate { x: -85.597, y: 225.618 }, Coordinate { x: -114.48, y: 359.999 }, Coordinate { x: -117.017, y: 360.0 }, Coordinate { x: -85.0, y: 215.0 }, Coordinate { x: -85.0, y: 155.0 }, Coordinate { x: -115.0, y: 155.0 }, Coordinate { x: -154.161, y: 360.0 }, Coordinate { x: -640.0, y: 360.0 }, Coordinate { x: -640.0, y: -360.0 }, Coordinate { x: 640.0, y: -360.0 }, Coordinate { x: 640.0, y: 360.0 }, Coordinate { x: 313.277, y: 360.0 }, Coordinate { x: 313.276, y: 359.999 },
```
visualization:
![image](https://user-images.githubusercontent.com/44753941/175775076-6bc0631a-abd6-4f08-9967-0eb3b75af503.png)

the second polygon painted red here:
![image](https://user-images.githubusercontent.com/44753941/175775197-b2b72c13-e6df-4ae1-bbb6-46dec0dba616.png)

I don't know if the polygon actually has a slit with a volume. I haven't scrutinized the data of the second polygon yet.

Some context: I'm doing game dev and implementing [visuability polygon](https://en.wikipedia.org/wiki/Visibility_polygon)  using the [geo-visibility](https://docs.rs/geo-visibility/latest/geo_visibility/) crate. The `geo-visibility` crate produces the inner irregular `Polygon` and I'm doing differencing with the screen in order to apply "fog of war".
![vis](https://user-images.githubusercontent.com/44753941/175775424-6a1c1603-9f89-4d9a-9390-cb38132654d2.gif)

**Comment by urschrei** [comment]
It looks like the debug assertion is being triggered because the two chains have different winding orders. Have you ensured that the polygons you're attempting to operate on have the same winding order?

**Comment by Madoshakalaka** [comment]
> It looks like the debug assertion is being triggered because the two chains have different winding orders. Have you ensured that the polygons you're attempting to operate on have the same winding order?

Is there anything I can use in `geo` to ensure their winding order?

**Comment by urschrei** [comment]
https://docs.rs/geo/0.22.0/geo/algorithm/orient/trait.Orient.html

**Comment by Madoshakalaka** [comment]
I tried orienting them, but it still panics at the same kind of debug assertion, my code:
```rust
    let screen = screen.orient(Direction::Default);
    let vis = vis.orient(Direction::Default);
    // trace!("Differencing, screen: {:?}, vis: {:?}", screen, vis);
    let fog = screen.difference(&vis);
```
panic message:
```
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `CounterClockwise`,
 right: `Clockwise`', /home/maa/.cargo/registry/src/github.com-1ecc6299db9ec823/geo-0.22.0/src/algorithm/bool_ops/rings.rs:159:13
```

**Comment by urschrei** [comment]
Have you also checked that your polygon is valid? OGC Simple Feature Polygons can't have "spikes" (only closed rings), and it looks like yours might (top centre):
<img width="822" alt="Screenshot 2022-06-25 at 17 35 16" src="https://user-images.githubusercontent.com/131862/175782700-d55f5d62-263f-4f00-ac75-9baec629f483.png">

**Comment by urschrei** [comment]
Actually nm, I scaled it down and it's definitely valid:
<img width="611" alt="Screenshot 2022-06-25 at 17 47 33" src="https://user-images.githubusercontent.com/131862/175783054-6f253d89-a850-4382-a5e7-b1c59733f614.png">

**Comment by rmanoka** [comment]
Thanks for reporting this @Madoshakalaka ; it is from an incorrect assumption in the part that constructs the output geometry.  Fixed in #869
